### PR TITLE
Coroutine

### PR DESCRIPTION
**Pull request by amusinamaria**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -1,7 +1,10 @@
 plugins {
     id 'com.android.application'
     id 'kotlin-android'
-    id 'org.jetbrains.kotlin.android.extensions'
+    id 'org.jetbrains.kotlin.plugin.serialization'
+    id 'kotlin-kapt'
+    id 'dagger.hilt.android.plugin'
```
Comment (Pavel-Sidorovich) [inline app/build.gradle:6]: For what?

File: app/src/main/java/com/github/amusinamaria/MoviesApp.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.github.amusinamaria
+
+import android.app.Application
+import dagger.hilt.android.HiltAndroidApp
+
+@HiltAndroidApp
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/MoviesApp.kt:6]: For what?

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+private suspend fun loadActors(context: Context): List<Actor> = withContext(Dispatchers.IO) {
+    val data = readAssetFileToString(context, "people.json")
+    parseActors(data)
+}
+
+internal fun parseActors(data: String): List<Actor> {
+    val jsonActors = jsonFormat.decodeFromString<List<Actor>>(data)
+    return jsonActors.map { Actor(id = it.id, name = it.name, profilePicture = it.profilePicture) }
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: You get a List of Actors from decodeFromString(). You don't need to convert

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+private val jsonFormat = Json { ignoreUnknownKeys = true }
+
+private suspend fun loadGenres(context: Context): List<Genre> = withContext(Dispatchers.IO) {
+    val data = readAssetFileToString(context, "genres.json")
+    parseGenres(data)
+}
+
+internal fun parseGenres(data: String): List<Genre> {
+    val jsonGenres = jsonFormat.decodeFromString<List<Genre>>(data)
+    return jsonGenres.map { Genre(id = it.id, name = it.name) }
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: You get a List of Genres from decodeFromString(). You don't need to convert

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    val jsonActors = jsonFormat.decodeFromString<List<Actor>>(data)
+    return jsonActors.map { Actor(id = it.id, name = it.name, profilePicture = it.profilePicture) }
+}
+
+@Suppress("unused")
+internal suspend fun loadMovies(context: Context): List<Movie> = withContext(Dispatchers.IO) {
+    val genresMap = loadGenres(context)
+    val actorsMap = loadActors(context)
+
+    val data = readAssetFileToString(context, "data.json")
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: This is synchronous code as far as I know.

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    genres: List<Genre>,
+    actors: List<Actor>
+): List<Movie> {
+    val genresMap = genres.associateBy { it.id }
+    val actorsMap = actors.associateBy { it.id }
+
+    val jsonMovies = jsonFormat.decodeFromString<List<JsonMovie>>(data)
+
+    return jsonMovies.map { jsonMovie ->
+        @Suppress("unused")
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: For what this annotation?

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        (Movie(
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: What is the first pair of parentheses for?

File: app/src/main/java/com/github/amusinamaria/repository/data/Movie.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package com.github.amusinamaria.repository.data
+
+import android.os.Parcelable
+import kotlinx.parcelize.Parcelize
+import kotlinx.serialization.Serializable
+
+@Parcelize
+@Serializable
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/Movie.kt]: Remove.

File: app/src/main/java/com/github/amusinamaria/MoviesApp.kt (hunk 2 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/MoviesApp.kt:6]: I'll use it after merge (in a branch with viewmodel)

File: app/build.gradle (hunk 1 above)
Comment (amusinamaria) [reply app/build.gradle:6]: Let it be here. I need it in branch master

File: app/build.gradle [hunk 9]
```diff
@@ -47,11 +47,23 @@ dependencies {
     implementation 'androidx.appcompat:appcompat:1.2.0'
     implementation 'com.google.android.material:material:1.2.1'
     implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
+    //RatingBar
```
Comment (n-balashov) [inline app/build.gradle:50]: maybe space between // and title?

File: app/src/main/java/com/github/amusinamaria/repository/data/Actor.kt [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import android.os.Parcelable
+import kotlinx.parcelize.Parcelize
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Parcelize
+@Serializable
+data class Actor(
+    val id: Int,
```
Comment (n-balashov) [inline app/src/main/java/com/github/amusinamaria/repository/data/Actor.kt]: Better to use Long

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 7 above)
Comment (n-balashov) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: to make it warmer

File: app/src/main/java/com/github/amusinamaria/ui/list/MoviesListFragment.kt [hunk 11]
```diff
@@ -8,14 +8,18 @@ import androidx.fragment.app.Fragment
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
 
 class MoviesListFragment : Fragment() {
 
     private var _binding: FragmentMoviesListBinding? = null
     private val binding get() = _binding!!
     private lateinit var moviesAdapter: MoviesAdapter
+    private val scope = CoroutineScope(Dispatchers.Main)
```
Comment (n-balashov) [inline app/src/main/java/com/github/amusinamaria/ui/list/MoviesListFragment.kt:23]: If you use this method you should call cancel() when a fragment is destroyed. Try to use LifecycleOwner.lifecycleScope

**Review by n-balashov — APPROVED** [review]

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 3 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Done

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 4 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Done

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 6 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Done

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 7 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Done

File: app/src/main/java/com/github/amusinamaria/repository/data/Movie.kt (hunk 8 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/Movie.kt]: Done

File: app/build.gradle (hunk 9 above)
Comment (amusinamaria) [reply app/build.gradle:50]: No

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 5 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Done?

File: app/src/main/java/com/github/amusinamaria/repository/data/Actor.kt (hunk 10 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/Actor.kt]: Why long?

**Review by nikslushkin — APPROVED** [review]

File: app/src/main/java/com/github/amusinamaria/repository/data/Actor.kt (hunk 10 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/Actor.kt]: Done

File: app/src/main/java/com/github/amusinamaria/ui/list/MoviesListFragment.kt (hunk 11 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/ui/list/MoviesListFragment.kt:23]: Done

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+import kotlinx.serialization.json.Json
+
+private val jsonFormat = Json { ignoreUnknownKeys = true }
+
+private suspend fun loadGenres(context: Context): List<Genre> = withContext(Dispatchers.IO) {
+    val data = readAssetFileToString(context, "genres.json")
+    parseGenres(data)
+}
+
+internal fun parseGenres(data: String): List<Genre> {
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Maybe private?

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    coroutineScope {
+        launch {
+            genresMap = loadGenres(context)
+        }
+        launch {
+            actorsMap = loadActors(context)
+        }
+    }
+
+    val data = readAssetFileToString(context, "data.json")
```
Comment (Pavel-Sidorovich) [inline app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt:48]: Maybe also to launch?

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 12 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt]: Done

File: app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt (hunk 13 above)
Comment (amusinamaria) [reply app/src/main/java/com/github/amusinamaria/repository/data/JsonLoad.kt:48]: Done

**Pull request merged**